test/publish.test.js

```diff
@@ -56,7 +56,7 @@ test.serial("Publish a release with templated path", async (t) => {
   const encodedGitTag = encodeURIComponent(nextRelease.gitTag);
   const generic = { path: "${env.FIXTURE}.txt", filepath: "/upload.txt" };
   const assets = [generic];
-  const uploaded = { url: "/uploads/upload.txt", alt: "upload.txt" };
+  const uploaded = getUploadResponse(generic.filepath, encodedRepoId);
```

Comment:
I would prefer to keep the test cases mostly self-contained and hardcode the upload responses like we do for most other test data.

Reply:
It is my opinion that using a helper function to mimic the gitlab api endpoint responses offers the following benefits over hardcoded values.

- Increased readability. It is more clear what determines the url values passed to the simulated release endpoint which are programatically generated by gitlab and not semantic-release it also removes the implication of some variables being treated as a gitlab response which are not actually contained in response data from the gitlab markdown upload endpoint. see [parameters expected in release request taken from variable uploaded not supplied by gitlab markdown upload endpoint](https://github.com/semantic-release/gitlab/pull/754/commits/35c76790fbe11d88f8c968c180a062964e562e21#diff-e5529dd6d95cc6412d41048a450c6d2b3647beaba2e5489a02094ab5b4441d4fL334-R336) and [uploaded implied to be a response from the gitlab markdown upload endpoint](https://github.com/semantic-release/gitlab/pull/754/commits/35c76790fbe11d88f8c968c180a062964e562e21#diff-e5529dd6d95cc6412d41048a450c6d2b3647beaba2e5489a02094ab5b4441d4fR342-R344)
- Increased maintainability. Should gitlab change their endpoint (again) updates need only to be made to [the mocked gitlab upload endpoint](https://github.com/semantic-release/gitlab/pull/754/commits/d4e057647f31bfb929f28a8c283ed96394aa9836#diff-9613b6e899c4d22ecbc1d582369aa824b4e77744ec607b5addacb9e65fb55f24R46-R58) instead of each test case being updated individually.
- Increased reliability. Although no test case currently exists, breakages potentially caused by the gitlab upload behavior of [only using the file name and ignoring the file path](https://github.com/semantic-release/gitlab/pull/754/commits/d4e057647f31bfb929f28a8c283ed96394aa9836#diff-9613b6e899c4d22ecbc1d582369aa824b4e77744ec607b5addacb9e65fb55f24R46-R48) would be more easily detected by mimicking the gitlab endpoint response than relying on hardcoded values and expecting developers to properly account for this, especially given that as far as I can tell this behavior is not actually documented anywhere (including in gitlab's documentation).

Reply:
[gitlab response handling in publish.js](https://github.com/semantic-release/gitlab/blob/d53f94c391792f81bc27528a428ca561cd8fb905/lib/publish.js#L146) would also contain the unexpected variables in point 1, which makes that point both a readability and reliability concern.

Reply:
Thanks for providing you reasoning @aldenbe, it's very clear!

I do agree with you in general, but seeing that we have many other similar places where we "hardcode" responses, e.g. https://github.com/semantic-release/gitlab/blob/4554f49a34e064fd26c8bd968f83841ee81ee914/test/publish.test.js#L181-L182 I would prefer to keep that style _in this PR_. I would be very open to accept a PR that refactors all the responses towards the style you suggest here. Would you be willing to give that a try?

@JonasSchubert WDYT?

Reply:
Simplified PR created explicitly addressing this issue created https://github.com/semantic-release/gitlab/pull/755

Unfortunately I have spent too much time on this already. Feel free to take anything desired from https://github.com/aldenbe/semantic-release-gitlab/tree/gitlab-mock-upload-endpoint and use it as you please.

Reply:
Thanks again @aldenbe for your work and fix.
I agree with @fgreinacher to keep the context of PRs to one topic to avoid too many changes in one. This usually helps to review a PR. I would be happy to see you create a second PR with your changes for the tests, @aldenbe. This can indeed improve readability. I understand if you have no time for this now, but if you would like to tackle that in the future it would be appreciated.